### PR TITLE
api: deprecated version annotation introduction

### PR DESCRIPTION
**Pull request by adisuissa**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: api: deprecated version annotation introduction
Additional Description:
Adding a deprecated API version annotation to deprecated fields and enum values in proto files.
This is part of the work on adding minor/patch versioning work.

Risk Level: Low (adding annotation to existing protos).
Testing: Added and modified tests for the tooling (in `tools/testdata`).
Docs Changes:
Release Notes:
Platform Specific Features: None

### CODE REVIEW

**Comment by adisuissa** [comment]
The first commit [3ffd898](https://github.com/envoyproxy/envoy/commit/3ffd8989f53d8419b2252eec2ae783f755519ffe) has the relevant scripts changes (including tests).
The second commit [8f9b997](https://github.com/envoyproxy/envoy/pull/15564/commits/8f9b997700795ef8f52b7336d83eaca8e7137f93) contains the updates to all protos in `api` and `generated_api_shadow`.

File: tools/api_versioning/generate_api_version_header_test.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 """Tests the api version header file generation.
 """
 import generate_api_version_header
-from generate_api_version_header import ApiVersion
+import utils
+from utils import ApiVersion
```
Comment (htuch) [inline tools/api_versioning/generate_api_version_header_test.py:5]: I think the `import utils` is redundant here and above.

File: tools/api_versioning/utils.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  same as the latest version, and the patch number is always 0. In addition,
+  the minor number is at least 0, and the oldest api version cannot be set
+  to a previous major number.
+
+  Args:
+    current_version: the current API version.
+
+  Returns:
+    the oldest supported API version.
+  """
```
Comment (htuch) [inline tools/api_versioning/utils.py:44]: I just merged https://github.com/envoyproxy/envoy/pull/15570 from @phlax, can you match style?

File: tools/protoxform/protoprint.py [hunk 3]
```diff
@@ -562,12 +573,43 @@ def format_reserved(enum_or_msg_proto):
...
+    if version.count('.') != 1:
+        return False
+    # Validate major and minor parts.
+    try:
+        major, minor = [int(x) for x in version.split('.')]
+        if major <= 0:
+            return False
+        if minor < 0:
+            return False
+    except:
```
Comment (htuch) [inline tools/protoxform/protoprint.py:595]: Wildcard `except` is a Python stylistic no-no. Can you be explicit on what exceptions?

File: tools/protoxform/protoprint.py [hunk 4]
```diff
@@ -580,6 +622,27 @@ def visit_service(self, service_proto, type_context):
...
+                        if not validate_deprecation_version(v.options.Extensions[
+                                deprecation_pb2.deprecated_at_minor_version_enum]):
+                            raise ProtoPrintError(
+                                'Error while parsing "deprecated_at_minor_version_enum" annotation value for enum value %s.'
+                                % v.name)
+                    else:
+                        self._needs_deprecation_annotation_import = True
+                        v.options.Extensions[
+                            deprecation_pb2.
+                            deprecated_at_minor_version_enum] = self._deprecated_annotation_version_value
```
Comment (htuch) [inline tools/protoxform/protoprint.py:645]: Suggest splitting into an independent method, the logic is complex enough.

File: tools/protoxform/protoprint.py [hunk 5]
```diff
@@ -580,6 +622,27 @@ def visit_service(self, service_proto, type_context):
...
+        # Verify that not hidden deprecated enum values of non-frozen protos have valid version
+        # annotations.
+        for v in enum_proto.value:
+            if v.options.deprecated and not self._frozen_proto and \
+                    not protoxform_options.has_hide_option(v.options):
+                # If the enum value has annotation from deprecation.proto, need to import it.
+                if v.options.HasExtension(deprecation_pb2.deprecated_at_minor_version_enum) or \
+                        v.options.HasExtension(deprecation_pb2.disallowed_by_default_enum):
+                    self._needs_deprecation_annotation_import = True
+                if v.name != 'DEPRECATED_AND_UNAVAILABLE_DO_NOT_USE':
```
Comment (htuch) [inline tools/protoxform/protoprint.py:634]: We should have a named constant for that.

**Review by htuch — CHANGES_REQUESTED** [review]
This is really great @adisuissa, matches exactly what we need. @phlax could you take a pass over this from a Python & `tools/` perspective? 
/wait

File: tools/protoxform/protoprint.py [hunk 6]
```diff
@@ -625,6 +688,21 @@ def visit_message(self, msg_proto, type_context, nested_msgs, nested_enums):
                 fields += '%soneof %s {\n%s%s' % (oneof_leading_comment, oneof_proto.name,
                                                   oneof_trailing_comment,
                                                   format_options(oneof_proto.options))
+            # Verify that deprecated fields (that are not hidden and are not of frozen protos)
+            # have a minor version annotation.
+            if field.options.deprecated and not self._frozen_proto and \
```
Comment (phlax) [inline tools/protoxform/protoprint.py:693]: this code/logic seems to be duplicated - i guess moving it to a method will fix that

**Review by phlax — CHANGES_REQUESTED** [review]
hi @adisuissa 

this PR will need a rebase as i have been making quite invasive changes to the py code - apologies for that

overall looks like a very welcome addition - im wondering if we can/should use the rst `.. deprecated::` directive - but i guess that is beyond the scope of this PR

File: tools/api_versioning/generate_api_version_header_test.py (hunk 1 above)
Comment (adisuissa) [reply tools/api_versioning/generate_api_version_header_test.py]: Done.

File: tools/api_versioning/utils.py (hunk 2 above)
Comment (adisuissa) [reply tools/api_versioning/utils.py:42]: Merged and matched style.
Must note that 4 spaces makes the lines much longer :)

File: tools/protoxform/protoprint.py (hunk 3 above)
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Fixed.

File: tools/protoxform/protoprint.py (hunk 4 above)
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Created a new function. Done.

File: tools/protoxform/protoprint.py (hunk 5 above)
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Done.

File: tools/api_versioning/utils.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    """Returns the API version from a given API version input file.
+
+  Args:
+    input_path: the file containing the API version (API_VERSION).
+
+  Returns:
+    a namedtuple containing the major, minor, patch versions.
+  """
+    lines = pathlib.Path(input_path).read_text().splitlines()
+    assert (len(lines) == 1)
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply tools/api_versioning/utils.py]: Removed.

File: tools/api_versioning/utils.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    # Mapping each field to int verifies it is a valid version
+    version = ApiVersion(*map(int, lines[0].split('.')))
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply tools/api_versioning/utils.py]: Removed.

File: tools/protoxform/protoprint.py [hunk 9]
```diff
@@ -33,7 +35,7 @@
 # this also serves as allowlist of extended options.
 from google.api import annotations_pb2 as _
 from validate import validate_pb2 as _
-from envoy.annotations import deprecation_pb2 as _
+from envoy.annotations import deprecation_pb2
 from envoy.annotations import resource_pb2
 from udpa.annotations import migrate_pb2
 from udpa.annotations import security_pb2 as _
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Thanks for the heads up.

File: tools/protoxform/protoprint.py (hunk 6 above)
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Thanks. It differed a bit, and now if a field is named `DEPRECATED_AND_UNAVAILABLE_DO_NOT_USE` (previously was only relevant for enum values) then it will also be ignored.

**Comment by adisuissa** [comment]
@htuch, @phlax thanks for the comments!
It should be all set now.

**Review by phlax — CHANGES_REQUESTED** [review]
thanks for update - comments inline

File: tools/protoxform/protoprint.py [hunk 10]
```diff
@@ -553,12 +568,78 @@ def format_reserved(enum_or_msg_proto):
...
+        _needs_deprecation_annotation_import flag is set.
+        The function also validates that if a field or enum value already have the deprecated
+        annotation value, then this value is a valid one ("X.Y" where X and Y are valid major,
+        and minor versions, respectively).
+        """
+        if field_or_evalue.options.deprecated and not self._frozen_proto and \
+                not protoxform_options.has_hide_option(field_or_evalue.options):
+            # If the field or enum value has annotation from deprecation.proto, need to import it.
+            if field_or_evalue.options.HasExtension(deprecation_tag) or \
+                    field_or_evalue.options.HasExtension(disallowed_tag):
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply tools/protoxform/protoprint.py:626]: This would then set/reset the flag on each invocation (not a flag that is only raised once).
I can do the following:
```
self._needs_deprecation_annotation_import = (
    self._needs_deprecation_annotation_import
    or field_or_evalue.options.HasExtension(deprecation_tag)
    or field_or_evalue.options.HasExtension(disallowed_tag))
```

Do you think it's more understandable this way?

[the remark this answers is not in the document]
Comment (phlax) [reply tools/protoxform/protoprint.py:626]: yep wfm

File: tools/protoxform/protoprint.py [hunk 11]
```diff
@@ -18,21 +19,25 @@
 import sys
 
 from tools.api_proto_plugin import annotations, traverse, visitor
+import tools.api_versioning.utils as api_version_utils
 from tools.protoxform import options as protoxform_options, utils
```
[the remark this answers is not in the document]
Comment (htuch) [reply tools/protoxform/protoprint.py:23]: @phlax can we have tooling to automate this? We have custom Python to do this for C++ to enforce Envoy opinionated include order.

[the remark this answers is not in the document]
Comment (phlax) [reply tools/protoxform/protoprint.py:23]: `isort` is the goto for this altho it might fight with yapf (i think i saw something about compatibility tho)

**Comment by htuch** [comment]
@adisuissa can you merge main? @phlax I think we should stop merging format fixes until this PR merges.

File: tools/protoxform/protoprint.py [hunk 12]
```diff
@@ -162,13 +167,16 @@ def format_type_context_comments(type_context, annotation_xforms=None):
...
+def format_header_from_file(source_code_info, file_proto, empty_file, import_deprecation_proto,
+                            frozen_proto):
     """Format proto header.
 
     Args:
         source_code_info: SourceCodeInfo object.
         file_proto: FileDescriptorProto for file.
         empty_file: are there no message/enum/service defs in file?
+        import_deprecation_proto: should the "envoy/annotation/deprecation.proto" be imported?
+        frozen_proto: is the proto file frozen?
```
Comment (htuch) [inline tools/protoxform/protoprint.py:179]: The logic here is a bit hard to follow; could we have something more descriptive, e.g. `has_minor_version_deprecation` and then infer these other properties?

File: tools/protoxform/protoprint.py [hunk 13]
```diff
@@ -553,12 +568,79 @@ def format_reserved(enum_or_msg_proto):
...
+    # Validate major and minor parts.
+    try:
+        major, minor = [int(x) for x in version.split('.')]
+        if major <= 0:
+            return False
+        if minor < 0:
+            return False
+    except ValueError:
+        return False
+    return True
```
Comment (htuch) [inline tools/protoxform/protoprint.py:592]: Should this belong in `utils`? Good candidate for unit tests.

**Review by htuch — CHANGES_REQUESTED** [review]
Looks good generally! Couple of minor comments..

**Comment by phlax** [comment]
> @phlax I think we should stop merging format fixes until this PR merges.

yep k - apologies @adisuissa - py formatting has been a bit of shifting target and been a merge/rebase nightmare

File: tools/api_versioning/utils.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    """Returns the API version from a given API version input file.
+
+    Args:
+      input_path: the file containing the API version (API_VERSION).
+
+    Returns:
+      a namedtuple containing the major, minor, patch versions.
+    """
+    lines = pathlib.Path(input_path).read_text().splitlines()
+    assert len(lines) == 1
```
[the remark this answers is not in the document]
Comment (phlax) [reply tools/api_versioning/utils.py:20]: also if this triggers it will raise an assertion errors about lengh which might not be that helpful

**Comment by phlax** [comment]
docs are rendered here:

https://storage.googleapis.com/envoy-pr/15564/docs/index.html

maybe i misunderstand what is included here - or the purpose - but i expected to find some change to the docs re deprecation

File: tools/api_versioning/utils.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (adisuissa) [reply tools/api_versioning/utils.py]: I think an assertion/exception is needed in case the API_VERSION file contents are accidentally modified.
I've added a comment to the assertion failure, but let me know if you think it should be removed.

File: tools/protoxform/protoprint.py (hunk 12 above)
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Done.

File: tools/protoxform/protoprint.py (hunk 13 above)
Comment (adisuissa) [reply tools/protoxform/protoprint.py]: Moved, and added tests.

**Comment by adisuissa** [comment]
> > @phlax I think we should stop merging format fixes until this PR merges.
> 
> yep k - apologies @adisuissa - py formatting has been a bit of shifting target and been a merge/rebase nightmare

No worries, I can fix the format if needed (it would just be easier to see the relevant changes if there were no formatting changes needed).

> docs are rendered here:
> 
> https://storage.googleapis.com/envoy-pr/15564/docs/index.html
> 
> maybe i misunderstand what is included here - or the purpose - but i expected to find some change to the docs re deprecation

My plan is to add an explanation to `api/API_VERSIONING.md` (see draft: https://github.com/envoyproxy/envoy/pull/14943).
I can also update the docs, but not sure where. It seems that the [API versioning](https://storage.googleapis.com/envoy-pr/1f5154d/docs/configuration/overview/versioning.html) points to the internal doc.
Another possibility is to add the explanation to [this FAQ](https://storage.googleapis.com/envoy-pr/1f5154d/docs/faq/configuration/deprecation.html?highlight=deprecated).

File: tools/api_versioning/utils.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (phlax) [reply tools/api_versioning/utils.py]: if you keep it then add an exception that will be useful when the code runs

with just an `assert` you would get something like this

```python
>>> assert len([]) == 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```

ie no useful info

i think you need to use an error that will provide info, eg:

```python
if len(lines) != 1:
    raise APIHeaderError("Explanation of what has gone wrong")
```

not sure exactly what kind of error or message is best

**Comment by phlax** [comment]
> My plan is to add an explanation to api/API_VERSIONING.md

if we have a way of saying "this feature is deprecated in version X" it would be really helpful to include this in the docs that end users read - im still not sure if i understand the purpose exactly - or whether this provides that

File: tools/api_versioning/utils.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (adisuissa) [reply tools/api_versioning/utils.py]: Hmmm, interesting...
 I was under the impression that assertion failure also prints the line number. This is what I get at the moment:
 ```
 Traceback (most recent call last):
  File "/home/adip/git/envoy/bazel-bin/tools/protoxform/protoprint.runfiles/envoy/tools/protoxform/protoprint.py", line 733, in <module>
    traverse.traverse_file(file_proto, ProtoFormatVisitor(api_version_file_path, frozen_proto)))
  File "/home/adip/git/envoy/bazel-bin/tools/protoxform/protoprint.runfiles/envoy/tools/protoxform/protoprint.py", line 586, in __init__
    current_api_version = api_version_utils.get_api_version(api_version_file_path)
  File "/home/adip/git/envoy/bazel-bin/tools/protoxform/protoprint.runfiles/envoy/tools/api_versioning/utils.py", line 28, in get_api_version
    assert len(lines) == 1, 'Api version file must have a single line of format X.Y.Z, where X, Y, and Z are non-negative integers.'
AssertionError: Api version file must have a single line of format X.Y.Z, where X, Y, and Z are non-negative integers.
```

I don't object using an exception, although there are other errors that may emit an assertion failure (e.g., empty lines), so I'll just add explicit handling of these.

[the remark this answers is not in the document]
Comment (phlax) [reply tools/api_versioning/utils.py]: yep - use a proper exception - its also catchable then

File: tools/api_versioning/utils.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def get_api_version(input_path):
+    """Returns the API version from a given API version input file.
+
+    Args:
+      input_path: the file containing the API version (API_VERSION).
+
+    Returns:
+      a namedtuple containing the major, minor, patch versions.
```
Comment (htuch) [inline tools/api_versioning/utils.py:22]: Nit: `Throws` docsstring

**Review by htuch — DISMISSED** [review]
LGTM, thanks! @phlax if you're good we can merge this.

File: tools/api_versioning/utils.py (hunk 15 above)
Comment (adisuissa) [reply tools/api_versioning/utils.py:22]: Added the `Raises` clause to the docstring.

**Comment by adisuissa** [comment]
> if we have a way of saying "this feature is deprecated in version X" it would be really helpful to include this in the docs that end users read - im still not sure if i understand the purpose exactly - or whether this provides that

I agree that the docs should be updated, once the patch version capabilities will be added (until now the work has mainly added minor version capabilities).
Moreover, we'll need to clarify which deprecated fields are no longer supported in a specific version.

@phlax thanks for the thorough review and great comments!

**Review by phlax — APPROVED** [review]
lgtm, thanks @adisuissa 

im looking forward to getting a better understanding of how we can use in docs
